spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -157,34 +170,38 @@ public DatastoreEntityConverter getDatastoreEntityConverter() {
 
 	@Override
 	public <T> void deleteById(Object id, Class<T> entityClass) {
-		getDatastoreReadWriter().delete(getKeyFromId(id, entityClass));
+		performDelete(new Key[] { getKeyFromId(id, entityClass) }, Collections.singletonList(id), null, entityClass);
 	}
 
 	@Override
 	public <T> void deleteAllById(Iterable<?> ids, Class<T> entityClass) {
-		List<Key> keys = getKeysFromIds(ids, entityClass);
-		getDatastoreReadWriter().delete(keys.toArray(new Key[keys.size()]));
+		performDelete(getKeysFromIds(ids, entityClass).toArray(new Key[0]), ids, null, entityClass);
 	}
 
 	@Override
 	public <T> void delete(T entity) {
-		getDatastoreReadWriter().delete(getKey(entity, false));
+		performDelete(new Key[] { getKey(entity, false) }, null, Collections.singletonList(entity), entity.getClass());
 	}
 
 	@Override
 	public <T> void deleteAll(Iterable<T> entities) {
-		getDatastoreReadWriter()
-				.delete(StreamSupport.stream(entities.spliterator(), false)
-				.map((x) -> getKey(x, false)).toArray(Key[]::new));
+		performDelete(StreamSupport.stream(entities.spliterator(), false)
+				.map((x) -> getKey(x, false)).toArray(Key[]::new), null, entities, null);
 	}
 
 	@Override
 	public long deleteAll(Class<?> entityClass) {
 		Key[] keysToDelete = findAllKeys(entityClass);
-		getDatastoreReadWriter().delete(keysToDelete);
+		performDelete(keysToDelete, null, null, entityClass);
 		return keysToDelete.length;
 	}
 
+	private void performDelete(Key[] keys, Iterable ids, Iterable entities, Class entityClass) {
```

Comment:
Requiring `Key[]` is causing you all these conversion from `List`.
Also, the only thing common across all deletes is the availability of the key and entity class.
Can we just provide those two pieces of information in the event for now?

Reply:
The conversion to Key[] needs to happen anyway because the client lib can only take the array (not the List). 

Does it hurt to have the IDs or java POJOs that were used to delete? Those can be very useful to have because the user can directly inspect the things being deleted which is either impossible or difficult if you only have the Key and entity type. Furthermore some types of delete (like by several POJOs where the types differ ) don't offer the entity class at all.

Reply:
The only issue with that is creating Iterables of ids or entities, even in case they will not be used. I think it can be useful to have them, but I'm not sure about those extra objects.

Reply:
What is your proposal ? @dmitry-s 

I wanted to avoid a zoo of delete events with both BEFORE and AFTER versions...

```
BeforeSingleEntityDeleteEvent
AfterSingleEntityDeleteEvent
BeforeMultiEntityDeleteEvent
AfterMultiEntityDeleteEvent
BeforeSingleIDDeleteEvent
AfterSingleIDDeleteEvent
BeforeByClassDeleteEvent
AfterByClassDeleteEvent
DeleteEventWithEntityClass
DeleteEventWithEntities
DeleteEventWithIDs
...
etc....
```

Reply:
From OOP perspective, you would probably have an event for each delete method if the fields vary in each. That being said, I think for now we can just make an event that only has the keys. We can always add others as needed that extend that event.

Reply:
@meltsufin  the problem with just holding keys is that the receiver of the events can't do anything with them. the user cannot retrieve any information using those  keys because of course the keys no longer exist in Datastore.